ImageDetection/ImageDetection/blob_detector.py

```diff
@@ -0,0 +1,132 @@
+
+        """
+        Set up filters. Values are changed in ImageDetectionSettings.
+        """
+        params.filterByArea = ImageDetectionSettings.FILTER_BY_AREA_ON
+        params.minArea = self.minimum_area
+        params.maxArea = self.maximum_area
+
+        params.filterByCircularity = ImageDetectionSettings.FILTER_BY_CIRCULARITY_ON
+        params.minCircularity = ImageDetectionSettings.MINIMUM_CIRCULARITY
+        params.maxCircularity = ImageDetectionSettings.MAXIMUM_CIRCULARITY
+
+        params.filterByConvexity = ImageDetectionSettings.FILTER_BY_CONVEXITY_ON
+        params.minConvexity = ImageDetectionSettings.MINIMUM_CONVEXITY
+        params.maxConvexity = ImageDetectionSettings.MAXIMUM_CONVEXITY
+
+        params.filterByInertia = ImageDetectionSettings.FILTER_BY_INERTIA_ON
+        params.minInertiaRatio = ImageDetectionSettings.MINIMUM_INERTIA_RATIO
+        params.maxInertiaRatio = ImageDetectionSettings.MAXIMUM_INERTIA_RATIO
+
+        """
+        Create a detector with the parameters.
+        """
+        ver = (cv2.__version__).split('.')
+        if int(ver[0]) < 3 :
+        	detector = cv2.SimpleBlobDetector(params)
+        else :
+        	detector = cv2.SimpleBlobDetector_create(params)
+
+        """
+        Detect blobs.
+        """
+        inverted_image = PIL.ImageOps.invert(self.raw_image)
+        posterized_image = PIL.ImageOps.posterize(inverted_image, 2)
+        image = cv2.cvtColor(numpy.array(posterized_image), cv2.COLOR_RGB2BGR)
+
+        keypoints = detector.detect(image)
+
+        """
+        Return the coordinates of the top left corners of the blobs and their
+        widths and heights with margins around it. The margins are checked so
+        that they do not extend out of the bound of the image.
+        """
+        blob_list = []
+        for index in range(len(keypoints)):
+            blob_center_x = round(keypoints[index].pt[0])
+            blob_center_y = round(keypoints[index].pt[1])
+            blob_diameter = round(keypoints[index].size)
+
+            blob_top_left_x = blob_center_x - blob_diameter
+            blob_top_left_y = blob_center_y - blob_diameter
+
+            blob_bottom_right_x = blob_center_x + blob_diameter
+            blob_bottom_right_y = blob_center_y + blob_diameter
+
+            if (blob_top_left_x < 0):
+                blob_top_left_x = 0
+
+            if (blob_top_left_y < 0):
+                blob_top_left_y = 0
+
+            if (blob_bottom_right_x >= self.raw_image.width):
+                blob_bottom_right_x = self.raw_image.width - 1
+
+            if (blob_bottom_right_y >= self.raw_image.height):
+                blob_bottom_right_y = self.raw_image.height - 1
+
+            blob_list.append([blob_top_left_x, blob_top_left_y, blob_bottom_right_x - blob_top_left_x, blob_bottom_right_y - blob_top_left_y])
+
+        """
+        Show the image with detected blobs circled.
+        """
```

Comment:
Single-line comments should begin with a `#` and not use the `"""` comment begin/close.